python/ray/tune/pbt.py

```diff
@@ -2,189 +2,248 @@
 
-    def on_trial_complete(self, trial_runner, trial, result):
-        self._results[trial].append(result)
-        self._completed_trials.add(trial)
-
-    def _exploit(self, hyperparameter_mutations, best_trial,
-                 trial, trial_runner, time):
-        trial.stop()
-        mutate_string = "_mutated@" + str(time)
-        hyperparams = copy.deepcopy(best_trial.config)
-        hyperparams = self._explore(hyperparams, hyperparameter_mutations,
-                                    best_trial)
-        print("new hyperparameter configuration: {0}".format(hyperparams))
-        checkpoint = self._checkpoint_paths[best_trial]
-        trial._checkpoint_path = checkpoint
-        trial.config = hyperparams
-        trial.experiment_tag = trial.experiment_tag + mutate_string
-        trial.start()
-
-    def _explore(self, hyperparams, hyperparameter_mutations, best_trial):
-        if hyperparameter_mutations is not None:
-            hyperparams = {
-                param: random.choice(hyperparameter_mutations[param])
-                for param in hyperparams
-                if param != "env" and param in hyperparameter_mutations
-            }
-            for param in best_trial.config:
-                if param not in hyperparameter_mutations and param != "env":
-                    hyperparams[param] = math.ceil(
-                        (best_trial.config[param]
-                         * random.choice([0.8, 1.2])/2.)) * 2
+    def _exploit(self, trial, trial_to_clone):
+        """Transfers perturbed state from trial_to_clone -> trial."""
+
+        trial_state = self._trial_state[trial]
+        new_state = self._trial_state[trial_to_clone]
+        if not new_state.last_checkpoint:
+            print("[pbt] warn: no checkpoint for trial, skip exploit", trial)
+            return
+        new_config = explore(
+            trial_to_clone.config, self._hyperparam_mutations,
+            self._resample_probability)
+        print(
+            "[exploit] transferring weights from trial "
+            "{} (score {}) -> {} (score {})".format(
+                trial_to_clone, new_state.last_score, trial,
+                trial_state.last_score))
+        # TODO(ekl) restarting the trial is expensive. We should implement a
+        # lighter way reset() method that can alter the trial config.
+        trial.stop(stop_logger=False)
+        trial.config = new_config
+        trial.experiment_tag = make_experiment_tag(
+            trial_state.orig_tag, new_config, self._hyperparam_mutations)
+        trial.start(new_state.last_checkpoint)
+        self._num_perturbations += 1
+
+    def _quantiles(self):
+        """Returns trials in the lower and upper `quantile` of the population.
+
+        If there is not enough data to compute this, returns empty lists."""
+
+        trials = []
+        for trial, state in self._trial_state.items():
+            if state.last_score is not None and not trial.is_finished():
```

Comment:
why not just return ranges? sort the last_score and then take the quartile points.

Reply:
Set semantics are slightly less error-prone, and here the performance doesn't matter so much so it's fine to take the hit